Source/WebGPU/WebGPU/PresentationContextIOSurface.mm

```diff
@@ -0,0 +1,137 @@
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "PresentationContextIOSurface.h"
+
+#import "APIConversions.h"
+#import "Adapter.h"
+
+// borrowed from pal/spi/cocoa/IOTypesSPI.h
+#if PLATFORM(MAC) || USE(APPLE_INTERNAL_SDK)
+#include <IOKit/IOTypes.h>
+#else
+
+enum {
+    kIOWriteCombineCache = 4,
+};
+
+enum {
+    kIOMapCacheShift = 8,
+    kIOMapWriteCombineCache = kIOWriteCombineCache << kIOMapCacheShift,
+};
+#endif
+
+namespace WebGPU {
+
+static NSDictionary *optionsFor32BitSurface(int width, int height, unsigned pixelFormat)
+{
+    unsigned bytesPerElement = 4;
+    unsigned bytesPerPixel = 4;
+
+    size_t bytesPerRow = IOSurfaceAlignProperty(kIOSurfaceBytesPerRow, width * bytesPerPixel);
+    ASSERT(bytesPerRow);
+
+    size_t totalBytes = IOSurfaceAlignProperty(kIOSurfaceAllocSize, height * bytesPerRow);
+    ASSERT(totalBytes);
+
+    return @{
+        (id)kIOSurfaceWidth: @(width),
+        (id)kIOSurfaceHeight: @(height),
+        (id)kIOSurfacePixelFormat: @(pixelFormat),
+        (id)kIOSurfaceBytesPerElement: @(bytesPerElement),
+        (id)kIOSurfaceBytesPerRow: @(bytesPerRow),
+        (id)kIOSurfaceAllocSize: @(totalBytes),
+#if PLATFORM(IOS_FAMILY)
+        (id)kIOSurfaceCacheMode: @(kIOMapWriteCombineCache),
+#endif
+        (id)kIOSurfaceElementHeight: @(1)
+    };
+}
+
+static RetainPtr<IOSurfaceRef> createIOSurface(int width, int height)
+{
+    NSDictionary *options = optionsFor32BitSurface(width, height, 'BGRA');
+    return adoptCF(IOSurfaceCreate((CFDictionaryRef)options));
+}
+
+static RetainPtr<IOSurfaceRef> createSurfaceFromDescriptor(const WGPUSurfaceDescriptor& descriptor)
+{
+    if (!descriptor.nextInChain || descriptor.nextInChain->sType != static_cast<WGPUSType>(WGPUSTypeExtended_SurfaceDescriptorCocoaSurfaceBacking))
+        return nullptr;
+
+    auto widthHeight = reinterpret_cast<const WGPUSurfaceDescriptorCocoaCustomSurface*>(descriptor.nextInChain);
+    return createIOSurface(widthHeight->width, widthHeight->height);
+}
+
+PresentationContextIOSurface::PresentationContextIOSurface(const WGPUSurfaceDescriptor& descriptor)
+    : m_displayBuffer(createSurfaceFromDescriptor(descriptor))
+    , m_drawingBuffer(createSurfaceFromDescriptor(descriptor))
+{
+}
+
+PresentationContextIOSurface::~PresentationContextIOSurface() = default;
+
+void PresentationContextIOSurface::configure(Device&, const WGPUSwapChainDescriptor&)
+{
+}
+
+void PresentationContextIOSurface::present()
+{
+    nextDrawable();
```

Comment:
this has become more of like `swapBuffers` instead of `nextDrawable` because we ignore the return value, but I think changing the name is outside the scope of this PR, we can address it in the future

Reply:
I'll keep it the way it is for now; I think we've got some larger refactoring to do here anyway in the near future.

Reply:
Sounds good 👍, seems reasonable to inline it at this point but certainly no rush to do so